feat: import and export of circuit files

pulkit-30 opened the pull request:
## [GSOC](https://github.com/CircuitVerse/CircuitVerse/wiki/GSoC%2722-Project-List#project-4---circuitverse-simulator-improvements)
 
Fixes #357 

#### Describe the changes you have made in this PR -
implemented the import and export of circuit files : A `.cv` extension file that contains the circuit data that can be exported and imported from/to the simulator


### Screenshots of the changes (If any) -


https://user-images.githubusercontent.com/76155456/178317879-d4ede84e-ff10-403d-9c41-0ed71750bbf5.mov

pulkit-30 commented on .eslintrc.json:
```diff
@@ -10,7 +10,7 @@
         "vars-on-top": 0,
         "no-unused-vars": 0,
-        "max-lines-per-function": 0
+        "max-lines-per-function": 0      
```
Nothing... [previous pr commits](https://github.com/CircuitVerse/CircuitVerse/pull/2243)

pulkit-30 replied:
sure 👍🏽

gr455 commented:
@pulkit-30 
1. Let's not change the title of the project title based on the filename
2. Can we also write on the file selection modal `Upload a file <upload button> or drag and drop here`? (also right now the feature works inconsistently; sometimes it does, sometimes it just uploads the file, but the load function is not triggered) 
3. Also thinking more about it, let's not be dependent on extensions. Validate the file before loading the circuit (validate that the json in the file is valid and contains everything we need) (I would like to know your thougthts on this, since this may not be a trivial task)

pulkit-30 commented:
>Also thinking more about it, let's not be dependent on extensions. Validate the file before loading the circuit (validate that the json in the file is valid and contains everything we need) (I would like to know your thougthts on this, since this may not be a trivial task)

@gr455  
worked on this, and can be easily implemented by following these simple steps-
1. store current scope.
2. try to load circuit data from file
3. if any error occurs... roll back and load stored data.
4. else load file data.

but here some questions arise
1. Currently we have a default `.cv` file extension, if we remove the dependency on the extension then which extension we should prefer? I think we should go with `.json`
2. we know that circuit data must be JSON, then why accept `.pdf`,  `.md`, `.txt` extensions?

pulkit-30 commented:
Also updated Import dialog box UI, added drag and drop feature 🚀 

<img width="670" alt="Screenshot 2022-07-24 at 6 18 11 PM" src="https://user-images.githubusercontent.com/76155456/180647688-0f48e650-69ff-4e71-944f-2b3a0bf9e514.png">

> also right now the feature works inconsistently; sometimes it does, sometimes it just uploads the file, but the load function is not triggered)

any specific error or situation? @gr455

gr455 commented:
> > Also thinking more about it, let's not be dependent on extensions. Validate the file before loading the circuit (validate that the json in the file is valid and contains everything we need) (I would like to know your thougthts on this, since this may not be a trivial task)
> 
> @gr455 worked on this, and can be easily implemented by following these simple steps-
> 
> 1. store current scope.
> 2. try to load circuit data from file
> 3. if any error occurs... roll back and load stored data.
> 4. else load file data.
> 

So this would be a fallback strategy which might not look very smooth. I was thinking if there is some way to validate from the file itself if it contains all the data we need (just think of it like a schema validation - just checking if the json has all the keys we need, the data itself maybe wrong but that would definitely beyond the scope of validation, and we could use this fallback stragtegy in that case)

We can at the end of the day use just the fallback strategy but that would be less seamless (in case some user uploads some other json)

> but here some questions arise
> 
> 1. Currently we have a default `.cv` file extension, if we remove the dependency on the extension then which extension we should prefer? I think we should go with `.json`
> 2. we know that circuit data must be JSON, then why accept `.pdf`,  `.md`, `.txt` extensions?

Just like linux doesn't care about extensions, we should also not. Even if a pdf file has a json file format, that's good enough for us. At the end of the day, extensions are really just filenames. We can stick to the .cv extension as a default when we export

gr455 commented:
> Also updated Import dialog box UI, added drag and drop feature rocket
> 
> <img alt="Screenshot 2022-07-24 at 6 18 11 PM" width="670" src="https://user-images.githubusercontent.com/76155456/180647688-0f48e650-69ff-4e71-944f-2b3a0bf9e514.png">
> 
> > also right now the feature works inconsistently; sometimes it does, sometimes it just uploads the file, but the load function is not triggered)
> 
> any specific error or situation? @gr455

Ahh, can't point out. Can't even reproduce now :disappointed:. I'll check some more

gr455 commented:
Also btw the UI looks good now!

pulkit-30 commented:
> So this would be a fallback strategy which might not look very smooth. I was thinking if there is some way to validate from the file itself if it contains all the data we need (just think of it like a schema validation - just checking if the json has all the keys we need, the data itself maybe wrong but that would definitely beyond the scope of validation, and we could use this fallback stragtegy in that case)
> 
> We can at the end of the day use just the fallback strategy but that would be less seamless (in case some user uploads some other json)

let me try 👍🏽

pulkit-30 commented on simulator/src/file/Open.js:
```diff
@@ -0,0 +1,66 @@
+    });
+    $('#ImportCircuitFilesDialog').focus();
+
+    function ValidateData(circuitData) {
```
validate function to validate JSON file data...

working:


https://user-images.githubusercontent.com/76155456/180723367-c3723aa5-e10e-47fa-9ab5-5723c4e110f9.mov

gr455 replied:
Ah great. Just one small nitpick: Can you think of a way to make the error message pop out a little more? Right now it's easy to miss

pulkit-30 replied:
<img width="680" alt="Screenshot 2022-07-26 at 12 45 58 AM" src="https://user-images.githubusercontent.com/76155456/180856766-245d056d-6e00-4f45-ab64-d777c894040d.png">

this seems good?

gr455 replied:
@pulkit-30 Sorry for a very late review. I would say not to bother the user with the exact error. Can we just make this a simple "Invalid file format" error?

I'll merge this PR (because it's complete and standing for too long). You can open another small PR for this change

pulkit-30 replied:
That's fine
I will do that in other PR 🚀 
Thanks 🤗

pulkit-30 commented on simulator/src/file/Open.js:
```diff
@@ -0,0 +1,66 @@
+        const valid = ValidateData(e.target.result); // validate data
+        if (!valid) {
+            // fallback
+            load(backUp);
```
fallback ... if any error occurs while loading file data... then it back up immediately.

tachyons commented on simulator/src/file/Open.js:
```diff
@@ -0,0 +1,66 @@
+
+    function ValidateData(fileData) {
+        try {
+            const parse_fileData = JSON.parse(fileData);
```
parsedFileDate
